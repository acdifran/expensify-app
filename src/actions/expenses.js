import moment from "moment";
import database from "../firebase/firebase";

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = moment()
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const expenseWithMillisTime = {
      ...expense,
      createdAt: createdAt.valueOf()
    };
    return database
      .ref("expenses")
      .push(expenseWithMillisTime)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

export const removeExpense = id => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = id => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return dispatch => {
    const updatesWithMillisTime = {
      ...updates,
      ...(!!updates.createdAt && { createdAt: updates.createdAt.valueOf() })
    };
    return database
      .ref(`expenses/${id}`)
      .update({ ...updatesWithMillisTime })
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref("expenses")
      .once("value")
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(expense => {
          expenses.push({
            id: expense.key,
            ...expense.val(),
            createdAt: moment(expense.val().createdAt)
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
