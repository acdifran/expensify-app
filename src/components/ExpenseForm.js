import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    const { expense = {} } = props;
    const {
      description = '',
      note = '',
      amount = '',
      createdAt = moment().valueOf()
    } = expense;
    this.state = {
      description,
      note,
      amount: typeof amount === 'number' ? amount / 100 : '',
      createdAt: moment(createdAt),
      calendarFocused: false,
      formError: undefined
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { description, note, amount, createdAt } = this.state;
    const { onSubmit } = this.props;
    if (!description || !amount) {
      this.setState(() => ({
        formError: 'Please provide description and amount'
      }));
    } else {
      this.setState(() => ({ formError: undefined }));
      onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt,
        note
      });
    }
  };

  render() {
    const {
      calendarFocused,
      formError,
      description,
      amount,
      createdAt,
      note
    } = this.state;
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {formError && <p className="form__error">{formError}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          value={description}
          onChange={this.onDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={this.onDateChange}
          focused={calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="text-area"
          placeholder="Add a note for your expense (optional)"
          value={note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button" type="submit">
            Save Expense
          </button>
        </div>
      </form>
    );
  }
}
