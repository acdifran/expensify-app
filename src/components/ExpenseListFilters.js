import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    const { setStartDate, setEndDate } = this.props;
    setStartDate(startDate);
    setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = e => {
    const { setTextFilter } = this.props;
    setTextFilter(e.target.value);
  };

  onSortChange = e => {
    const { sortByAmount, sortByDate } = this.props;
    if (e.target.value === 'amount') {
      sortByAmount();
    } else if (e.target.value === 'date') {
      sortByDate();
    }
  };

  render() {
    const { filters } = this.props;
    const { calendarFocused } = this.state;
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search Expenses"
              type="text"
              value={filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={filters.startDate}
              endDate={filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({ filters: state.filters });

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: date => dispatch(setStartDate(date)),
  setEndDate: date => dispatch(setEndDate(date))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
