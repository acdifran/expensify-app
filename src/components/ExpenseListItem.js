import React from "react";
import { Link } from "react-router-dom";
import numeral from "numeral";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format("$0,0.00")}-
      {createdAt.format("MMM Do, YYYY")}
    </p>
  </div>
);

export default ExpenseListItem;
