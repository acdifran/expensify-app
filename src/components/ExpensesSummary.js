import React from "react";
import numeral from "numeral";
import selectExpenses from "../selectors/selectExpenses";
import expensesTotal from "../selectors/expensesTotal";
import { connect } from "react-redux";

export const ExpensesSummary = props => {
  const expenseWord = props.expenses.length === 1 ? "expense" : "expenses";
  const totalExpenseValue = numeral(expensesTotal(props.expenses) / 100).format(
    "$0,0.00"
  );
  return (
    <div>
      {props.expenses.length > 0 && (
        <h3>
          Showing {props.expenses.length} {expenseWord} for {totalExpenseValue}
        </h3>
      )}
    </div>
  );
};

const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);
