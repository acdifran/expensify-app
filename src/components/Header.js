import React from "react";
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";

export default () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/dashboard" activeClassName="is-active" exact={true}>
      Home
    </NavLink>
    <NavLink to="/create" activeClassName="is-active" exact={true}>
      Create Expense
    </NavLink>
    <NavLink to="/help" activeClassName="is-active" exact={true}>
      Help
    </NavLink>
    <button onClick={startLogout()}>Logout</button>
  </header>
);
