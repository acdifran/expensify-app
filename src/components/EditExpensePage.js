import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = newExpense => {
    const { startEditExpense, history, expense } = this.props;
    startEditExpense(expense.id, newExpense);
    history.push('/');
  };

  onRemove = () => {
    const { startRemoveExpense, history, expense } = this.props;
    startRemoveExpense(expense.id);
    history.push('/');
  };

  render() {
    const { expense } = this.props;
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={expense} onSubmit={this.onSubmit} />
          <button
            className="button button--secondary"
            type="button"
            onClick={this.onRemove}
          >
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
