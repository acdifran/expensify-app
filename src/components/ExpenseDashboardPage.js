import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from "./ExpensesSummary";

export default () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
