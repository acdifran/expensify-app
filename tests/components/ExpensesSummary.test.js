import React from "react";
import { ExpensesSummary } from "../../src/components/ExpensesSummary";
import expenses from "../fixtures/expenses";
import { shallow } from "enzyme";

let wrapper;

beforeEach(() => {
  wrapper = shallow(<ExpensesSummary expenses={[]} />);
});

test("should render properly for no expenses", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render properly for one expense", () => {
  wrapper.setProps({ expenses: [expenses[0]] });
  expect(wrapper).toMatchSnapshot();
});

test("should render properly for multiple expenses", () => {
  wrapper.setProps({ expenses });
  expect(wrapper).toMatchSnapshot();
});
