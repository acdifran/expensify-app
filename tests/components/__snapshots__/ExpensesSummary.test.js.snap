// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly for multiple expenses 1`] = `
<div>
  <h3>
    Showing 
    3
     
    expenses
     for 
    $110.10
  </h3>
</div>
`;

exports[`should render properly for no expenses 1`] = `<div />`;

exports[`should render properly for one expense 1`] = `
<div>
  <h3>
    Showing 
    1
     
    expense
     for 
    $100.00
  </h3>
</div>
`;
